### PR TITLE
feat(linter): add dependency-checks to buildable lib generator

### PR DESCRIPTION
**Pull request by meeroslav**
This PR adds automatic configuration for `dependency-checks` rule to newly generated buildable libraries.

cc @jaysoo @leosvelperez 

## Related Issue(s)

Fixes #

### CODE REVIEW

File: docs/generated/packages/eslint-plugin/documents/dependency-checks.md [hunk 1]
```diff
@@ -8,7 +8,11 @@ We use the version numbers of the installed packages when checking whether the v
...
 ## Usage
 
-You can use the `dependency-checks` rule by adding it to your ESLint rules configuration:
+Library generators from `@nx` packages will configure this rule automatically when you opt-in for bundler/build setup.
+
+### Manual setup
+
+To set it up manually for existing libraries, you need to add the `dependency-checks` rule to your project's ESLint configuration:
 
 ```jsonc {% fileName=".eslintrc.json" %}
```
Comment (juristr) [inline docs/generated/packages/eslint-plugin/documents/dependency-checks.md:17]: should we add here 

`{% fileName="<your-project-root>/.eslintrc.json" %}`

Just to make it more clear this should be in your project rather than the root one?

File: docs/generated/packages/eslint-plugin/documents/dependency-checks.md [hunk 2]
```diff
@@ -26,7 +30,7 @@ You can use the `dependency-checks` rule by adding it to your ESLint rules confi
 }
 ```
 
-Linting `JSON` files is not enabled by default, so you will also need to add `package.json` to the `lintFilePatterns`:
+Additionally, you need to add the `package.json` to the `lintFilePatterns`:
```
Comment (juristr) [inline docs/generated/packages/eslint-plugin/documents/dependency-checks.md:33]: Optional, but I'd slightly rewrite to:

"Additionally, you need to adjust your `lintFilePatterns` to include the project's `package.json` file:"

File: docs/generated/packages/eslint-plugin/documents/dependency-checks.md [hunk 3]
```diff
@@ -8,7 +8,11 @@ We use the version numbers of the installed packages when checking whether the v
 
 ## Usage
```
Comment (juristr) [inline docs/generated/packages/eslint-plugin/documents/dependency-checks.md:9]: This is for the entire document. I'd have headings like

- Setup (in there mention that it should already be set up automatically, but here's how you do it manually
- Define the Build Target (or something similar; to me it wasn't obvious that the lint rule depends on a build target..like why would it. So I'd emphasize that by having a dedicated section here so ppl would immediately spot it. I'm even wondering whether the generated setup should include that rather than silently assuming `build` as the default 🤔, just to make it obvious )
- Overriding defaults (mention all the possible defaults one could override and how, basically just as you did there in the last paragraphs)

**Review by juristr — CHANGES_REQUESTED** [review]

**Review by juristr — APPROVED** [review]
LGTM, thx for adjusting

**Review by bcabanes — APPROVED** [review]

File: packages/plugin/src/generators/generator/generator.ts [hunk 4]
```diff
@@ -140,6 +146,22 @@ async function updateGeneratorJson(host: Tree, options: NormalizedSchema) {
       options.skipFormat
     );
   }
+  // add dependencies
+  updateJson<PackageJson>(
```
Comment (jaysoo) [inline packages/plugin/src/generators/generator/generator.ts:150]: Wouldn't nx and devkit already exist if users have `@nx/plugin` installed? And SWC packages may not be needed if they use tsc or another bunder.

**Review by jaysoo — APPROVED** [review]

**Comment by JamesHenry** [comment]
Looks good but Jack's question above seems reasonable...

File: packages/plugin/src/generators/generator/generator.ts (hunk 4 above)
Comment (meeroslav) [reply packages/plugin/src/generators/generator/generator.ts:150]: - They would be installed but as a transitive dependency. Our generated `generator` uses them directly so we need to reference them. Not to mention, using transitive deps does not work with `PnP`, `pnpm` (under some conditions), or the `no-hoist` flag.
- This is due to what our generator generates. The purpose is to not have broken `lint` after running the generator. 

If you look at our E2E test, it calls the following commands:
- nx generate @nx/plugin:plugin ${plugin} --linter=eslint
- nx generate @nx/plugin:generator ${generator} --project=${plugin}
- nx lint ${plugin}

Despite `compiler` being set by default to `tsc`, our helper functions check returns `swc`. Perhaps there's a bug elsewhere. I will look into this on Monday to be sure.

Comment (meeroslav) [reply packages/plugin/src/generators/generator/generator.ts:150]: The previously generated package.json was broken so it falsy requested `@swc`. After last commit from @jaysoo  the tests started failing as expected.

Comment (meeroslav) [reply packages/plugin/src/generators/generator/generator.ts:145]: Should be fixed now

**Review by jaysoo — APPROVED** [review]

File: e2e/expo/src/expo.test.ts [hunk 5]
```diff
@@ -22,6 +23,16 @@ describe('expo', () => {
...
+    // we create empty preset above which skips creation of `production` named input
+    updateJson('nx.json', (nxJson) => {
+      nxJson.namedInputs = {
+        default: ['{projectRoot}/**/*', 'sharedGlobals'],
+        production: ['default'],
+        sharedGlobals: [],
+      };
+      nxJson.targetDefaults.build.inputs = ['production', '^production'];
+      return nxJson;
+    });
```
Comment (FrozenPandaz) [inline e2e/expo/src/expo.test.ts:35]: Hm this seems wrong...

So if I create an `empty` workspace, then create an Angular app, the inputs aren't setup the same way as if I just create an Angular workspace?

File: packages/jest/src/generators/init/init.ts [hunk 6]
```diff
@@ -126,7 +126,9 @@ function addTestInputs(tree: Tree) {
       // Remove jest.config.js/ts
       '!{projectRoot}/jest.config.[jt]s',
       // Remove test-setup.js/ts
-      '!{projectRoot}/src/test-setup.[jt]s'
+      // TODO(meeroslav) this should be standardized
+      '!{projectRoot}/src/test-setup.[jt]s',
+      '!{projectRoot}/test-setup.[jt]s'
```
Comment (FrozenPandaz) [inline packages/jest/src/generators/init/init.ts:131]: Should we change this to `'!{projectRoot}/**/test-setup.[jt]s',`?

File: packages/js/src/generators/library/library.ts [hunk 7]
```diff
@@ -249,6 +255,25 @@ export async function addLint(
...
+    if (options.bundler === 'tsc') {
+      json.dependencies = {
+        ...json.dependencies,
+        tslib: tsLibVersion,
+      };
+    } else if (options.bundler === 'swc') {
+      json.dependencies = {
+        ...json.dependencies,
+        '@swc-node/register': swcNodeVersion,
+        '@swc/core': swcCoreVersion,
```
Comment (FrozenPandaz) [inline packages/js/src/generators/library/library.ts:269]: I thought `@swc/helpers` was the only one needed? It's the only one we add when we generate the package.json. Why are these necessary here?

File: packages/plugin/src/generators/executor/executor.ts [hunk 8]
```diff
@@ -16,6 +16,7 @@ import type { Schema } from './schema';
 import * as path from 'path';
 import { PackageJson } from 'nx/src/utils/package-json';
 import pluginLintCheckGenerator from '../lint-checks/generator';
+import { nxVersion } from '@nx/js/src/utils/versions';
```
Comment (FrozenPandaz) [inline packages/plugin/src/generators/executor/executor.ts:19]: I think this is coming from the wrong place. Is there no `nxVersion` here in this package? We can add one.

File: packages/plugin/src/generators/generator/generator.ts [hunk 9]
```diff
@@ -17,6 +17,7 @@ import * as path from 'path';
 import { hasGenerator } from '../../utils/has-generator';
 import pluginLintCheckGenerator from '../lint-checks/generator';
 import type { Schema } from './schema';
+import { nxVersion } from '@nx/js/src/utils/versions';
```
Comment (FrozenPandaz) [inline packages/plugin/src/generators/generator/generator.ts:20]: Again, this seems like the wrong place.

File: packages/plugin/src/generators/migration/migration.ts [hunk 10]
```diff
@@ -15,6 +15,7 @@ import type { Schema } from './schema';
 import * as path from 'path';
 import { addMigrationJsonChecks } from '../lint-checks/generator';
 import { PackageJson, readNxMigrateConfig } from 'nx/src/utils/package-json';
+import { nxVersion } from '@nx/js/src/utils/versions';
```
Comment (FrozenPandaz) [inline packages/plugin/src/generators/migration/migration.ts:18]: Again, this seems like the wrong place.

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

File: packages/plugin/src/generators/executor/executor.ts [hunk 11]
```diff
@@ -122,6 +123,18 @@ async function updateExecutorJson(host: Tree, options: NormalizedSchema) {
...
     );
   }
+  // add dependencies
+  updateJson<PackageJson>(
+    host,
+    joinPathFragments(options.projectRoot, 'package.json'),
+    (json) => {
+      json.dependencies = {
+        ...json.dependencies,
+        '@nx/devkit': nxVersion,
```
Comment (AgentEnder) [inline packages/plugin/src/generators/executor/executor.ts:133]: Can we swap the order of these? Or honestly, it may be better to do something like:

```
json.dependencies ??= {};
json.dependencies['@nx/devkit'] ??= readPackageVersionInRootPackageJson(tree, '@nx/devkit')
```

File: e2e/expo/src/expo.test.ts (hunk 5 above)
Comment (meeroslav) [reply e2e/expo/src/expo.test.ts:35]: This is the block where it's set https://github.com/nrwl/nx/blob/master/packages/workspace/src/generators/new/generate-workspace-files.ts#L116-L127

The `angular` has it only as a part of `ng-add` which is external command. All the `init` methods called by generators have something like this https://github.com/nrwl/nx/blob/master/packages/jest/src/generators/init/init.ts#L118-L119

File: packages/plugin/src/generators/executor/executor.ts (hunk 8 above)
Comment (meeroslav) [reply packages/plugin/src/generators/executor/executor.ts:19]: I will add one then.

File: packages/plugin/src/generators/executor/executor.ts (hunk 11 above)
Comment (meeroslav) [reply packages/plugin/src/generators/executor/executor.ts:133]: I have swapped the order (which might make sense if the user added `@nx/devkit`).

As for the second, I think users will anyway want to adapt their range, so I'd rather not assume anything. Using a fixed version is straightforward and something most people have in their root `package.json` anyway.

File: packages/plugin/src/generators/plugin/plugin.ts [hunk 12]
```diff
@@ -90,6 +98,32 @@ export async function pluginGenerator(host: Tree, schema: Schema) {
...
+      (json) => {
+        json.dependencies = {
+          ...json.dependencies,
+          '@swc-node/register': swcNodeVersion,
+          '@swc/core': swcCoreVersion,
+          '@swc/helpers': swcHelpersVersion,
+        };
+        return json;
+      }
+    );
```
[the remark this answers is not in the document]
Comment (meeroslav) [reply packages/plugin/src/generators/plugin/plugin.ts:126]: Yup, that exists in the js library generator. I'll remove it from here.

File: e2e/expo/src/expo.test.ts (hunk 5 above)
Comment (meeroslav) [reply e2e/expo/src/expo.test.ts:35]: if production named input wasn't created during the CNW then none of the generators would add it.

File: packages/js/src/generators/library/library.ts (hunk 7 above)
Comment (meeroslav) [reply packages/js/src/generators/library/library.ts]: Removed, that changed after @jaysoo's PR

File: packages/plugin/src/generators/plugin/plugin.ts [hunk 13]
```diff
@@ -21,6 +22,13 @@ import { addTsLibDependencies } from '@nx/js/src/utils/typescript/add-tslib-depe
 import { addSwcRegisterDependencies } from '@nx/js/src/utils/swc/add-swc-dependencies';
 
 import type { Schema } from './schema';
+import { type PackageJson } from 'nx/src/utils/package-json';
+import {
+  swcCoreVersion,
+  swcHelpersVersion,
+  swcNodeVersion,
+  tsLibVersion,
+} from '@nx/js/src/utils/versions';
```
Comment (FrozenPandaz) [inline packages/plugin/src/generators/plugin/plugin.ts:31]: This isn't needed, can you clean this up after?

**Review by FrozenPandaz — APPROVED** [review]
